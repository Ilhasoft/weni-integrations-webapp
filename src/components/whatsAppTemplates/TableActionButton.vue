<template>
  <unnnic-dropdown class="table-action__dropdown" :position="position">
    <unnnic-button
      slot="trigger"
      size="small"
      type="tertiary"
      iconCenter="navigation-menu-vertical-1"
    />

    <unnnic-dropdown-item
      :class="['table-action__dropdown__item', `table-action__dropdown__item--${option.id}`]"
      v-for="option in options"
      :key="option.id"
      @click="() => option.action()"
      :id="option.id"
    >
      <unnnic-icon-svg :icon="option.icon" size="sm" :scheme="option.scheme" />
      {{ option.title }}
    </unnnic-dropdown-item>
  </unnnic-dropdown>
</template>

<script>
  import { mapActions, mapState } from 'vuex';
  import { unnnicCallAlert } from '@weni/unnnic-system';

  export default {
    name: 'TableActionButton',
    props: {
      templateUuid: {
        type: String,
        required: true,
      },
      position: {
        type: String,
        default: 'bottom-left',
        validator(value) {
          return ['bottom-left', 'bottom-right', 'top-left', 'top-right'].indexOf(value) !== -1;
        },
      },
      data: {
        type: Object,
        required: true,
      },
    },
    data() {
      return {
        options: [
          {
            id: 'add_language',
            title: this.$t('WhatsApp.templates.table.actions.add_language'),
            icon: 'add-circle-1',
            scheme: 'neutral-darkest',
            action: () => {
              const { appCode, appUuid } = this.$route.params;
              this.$router.push({
                path: `/apps/my/${appCode}/${appUuid}/templates/edit/${this.templateUuid}`,
              });
            },
          },
          {
            id: 'see_details',
            title: 'See Details',
            icon: 'pencil-write-1',
            scheme: 'neutral-darkest',
            action: () => {
              this.setSelectedTemplate({ template: this.data });
              this.$router.push({
                path: `/apps/template-details`,
                hash: `#${this.data.uuid}`,
              });
            },
          },
          {
            id: 'delete_language',
            title: this.$t('WhatsApp.templates.table.actions.delete_language'),
            icon: 'bin-1-1',
            scheme: 'feedback-red',
            action: /* istanbul ignore next */ async () => {
              const { appUuid } = this.$route.params;
              await this.deleteTemplateMessage({ appUuid, templateUuid: this.templateUuid });

              if (this.errorDeleteTemplateMessage) {
                let errorMsg = this.$t('WhatsApp.templates.error.delete_template');
                const responseError = this.errorDeleteTemplateMessage.response.data.error;
                if (responseError) {
                  errorMsg = this.$t(responseError);
                }
                unnnicCallAlert({
                  props: {
                    text: errorMsg,
                    title: this.$t('general.error'),
                    icon: 'alert-circle-1-1',
                    scheme: 'feedback-red',
                    position: 'bottom-right',
                    closeText: this.$t('general.Close'),
                  },
                  seconds: 6,
                });
                return;
              }

              this.$emit('refresh-table');
            },
          },
        ],
      };
    },
    methods: {
      ...mapActions('WhatsApp', ['deleteTemplateMessage']),
      ...mapActions(['setSelectedTemplate']),
    },
    computed: {
      ...mapState('WhatsApp', ['errorDeleteTemplateMessage']),
    },
  };
</script>

<style lang="scss" scoped>
  .table-action {
    &__dropdown {
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      line-height: $unnnic-line-height-md + $unnnic-font-size-body-md;
      color: $unnnic-color-neutral-dark;

      ::v-deep .unnnic-dropdown__content {
        width: max-content;
      }

      &__item {
        &--delete_language {
          color: $unnnic-color-feedback-red;
        }
      }
    }
  }
</style>
