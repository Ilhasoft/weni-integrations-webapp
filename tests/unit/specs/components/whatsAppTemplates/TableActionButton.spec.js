import Vuex from 'vuex';
import VueRouter from 'vue-router';
import { mount, createLocalVue } from '@vue/test-utils';
import TableActionButton from '@/components/whatsAppTemplates/TableActionButton.vue';
import { unnnicDropdownItem, unnnicButton } from '@weni/unnnic-system';

const localVue = createLocalVue();
localVue.use(Vuex);
const router = new VueRouter();

describe('components/whatsAppTemplates/TableActionButton.vue', () => {
  let wrapper;
  let actions;
  let store;
  beforeEach(() => {
    actions = {
      setSelectedTemplate: jest.fn(),
      setAppUuid: jest.fn(),
    };
    store = new Vuex.Store({
      modules: {
        insights: {
          namespaced: true,
          actions: {
            setSelectedTemplate: jest.fn(),
            setAppUuid: jest.fn(),
          },
        },
      },
      actions,
    });
    wrapper = mount(TableActionButton, {
      localVue,
      router,
      store,
      mocks: {
        $t: () => 'some specific text',
        $route: {
          params: {
            appCode: 'code',
            appUuid: '123',
          },
        },
        $router: {
          push: jest.fn(),
        },
      },
      propsData: {
        templateUuid: '123',
        data: {
          uuid: 'uuid123',
        },
      },
    });
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render options properly', async () => {
    const triggerButton = wrapper.findComponent(unnnicButton);
    await triggerButton.trigger('click');
    expect(wrapper).toMatchSnapshot();
  });

  it('should go to edit template on Add Language option click', async () => {
    const triggerButton = wrapper.findComponent(unnnicButton);
    await triggerButton.trigger('click');
    const addLanguageOption = wrapper.findAllComponents(unnnicDropdownItem).at(0);

    expect(wrapper.vm.$router.push).not.toHaveBeenCalled();
    await addLanguageOption.trigger('click');
    expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      path: '/apps/my/code/123/templates/edit/123',
    });
  });

  it('should go to template details on See Details option click', async () => {
    const triggerButton = wrapper.findComponent(unnnicButton);
    await triggerButton.trigger('click');
    const seeDetails = wrapper.findAllComponents(unnnicDropdownItem).at(1);
    expect(seeDetails.exists()).toBe(true);

    expect(wrapper.vm.$router.push).not.toHaveBeenCalled();
    await seeDetails.trigger('click');
    expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1);
  });

  it('should call setSelectedTemplate', async () => {
    const triggerButton = wrapper.findComponent(unnnicButton);
    await triggerButton.trigger('click');
    const seeDetails = wrapper.findAllComponents(unnnicDropdownItem).at(1);
    expect(seeDetails.exists()).toBe(true);

    const spy = spyOn(wrapper.vm, 'setSelectedTemplate');
    expect(spy).not.toHaveBeenCalled();
    await seeDetails.trigger('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should call setAppUuid', async () => {
    const triggerButton = wrapper.findComponent(unnnicButton);
    await triggerButton.trigger('click');
    const seeDetails = wrapper.findAllComponents(unnnicDropdownItem).at(1);
    expect(seeDetails.exists()).toBe(true);

    const spy = spyOn(wrapper.vm, 'setAppUuid');
    expect(spy).not.toHaveBeenCalled();
    await seeDetails.trigger('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
