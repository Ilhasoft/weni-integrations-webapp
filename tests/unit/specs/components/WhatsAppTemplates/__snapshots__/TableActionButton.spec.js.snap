// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/WhatsAppTemplates/TableActionButton.vue should be rendered properly 1`] = `
<unnnicdropdown-stub class="table-action__dropdown">
  <unnnicbutton-stub size="small" type="terciary" iconcenter="navigation-menu-vertical-1"></unnnicbutton-stub>
  <unnnicdropdownitem-stub class="table-action__dropdown__item table-action__dropdown__item--add_language">
    <unnniciconsvg-stub icon="add-circle-1" size="sm" scheme="neutral-darkest"></unnniciconsvg-stub>
    some specific text
  </unnnicdropdownitem-stub>
  <unnnicdropdownitem-stub class="table-action__dropdown__item table-action__dropdown__item--duplicate_language">
    <unnniciconsvg-stub icon="copy-paste-1" size="sm" scheme="neutral-darkest"></unnniciconsvg-stub>
    some specific text
  </unnnicdropdownitem-stub>
  <unnnicdropdownitem-stub class="table-action__dropdown__item table-action__dropdown__item--delete_language">
    <unnniciconsvg-stub icon="bin-1-1" size="sm" scheme="feedback-red"></unnniciconsvg-stub>
    some specific text
  </unnnicdropdownitem-stub>
</unnnicdropdown-stub>
`;
